In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
cc_df = pd.read_csv('fraud.csv')

In [3]:
pd.set_option('display.max_columns', 50)
pd.set_option('display.max_rows', 50)

# 1. 데이터 둘러보기

- 불분명한 컬럼명 확인
    * amt:거래금액
    * city_pop: 도시 인구
    * dob: 생일
- 개인정보 관련 컬럼 삭제 필요
- 결측치 없으나 is_fraud 컬럼에 0/1 이외 값 유무 확인 필요
- 각 컬럼 타입은 큰 이상 없음
- 각 컬럼 값의 형식/범위가 맞는지 확인 필요

In [4]:
cc_df.shape

(491134, 22)

In [5]:
cc_df.head(3)

trans_date_trans_time               cc_num  \
0   2019-01-01 00:00:44         630423337322   
1   2019-01-01 00:12:34  4956828990005111019   
2   2019-01-01 00:17:16      180048185037117   

                            merchant     category     amt      first  \
0    fraud_Heller, Gutmann and Zieme  grocery_pos  107.23  Stephanie   
1  fraud_Schultz, Simonis and Little  grocery_pos   44.71    Kenneth   
2                  fraud_Kling-Grant  grocery_net   46.28       Mary   

       last gender                        street        city state    zip  \
0      Gill      F  43039 Riley Greens Suite 393      Orient    WA  99160   
1  Robinson      M            269 Sanchez Rapids   Elizabeth    NJ   7208   
2      Wall      F               2481 Mills Lock  Plainfield    NJ   7060   

       lat      long  city_pop                                job         dob  \
0  48.8878 -118.2105       149  Special educational needs teacher  1978-06-21   
1  40.6747  -74.2239    124967             Operational researcher  1980-12-21   
2  40.6152  -74.4150     71485             Leisure centre manager  1974-07-19   

                          trans_num   unix_time  merch_lat  merch_long  \
0  1f76529f8574734946361c461b024d99  1325376044  49.159047 -118.186462   
1  09eff9c806365e2a6be12c1bbab3d70e  1325376754  40.079588  -74.848087   
2  19e23c6a300c774354417befe4f31f8c  1325377036  40.021888  -74.228188   

   is_fraud  
0         0  
1         0  
2         0

In [6]:
cc_df.tail(3)

trans_date_trans_time            cc_num              merchant  \
491131   2020-12-31 23:59:09  3556613125071656  fraud_Hoppe-Parisian   
491132   2020-12-31 23:59:15  6011724471098086       fraud_Rau-Robel   
491133   2020-12-31 23:59:34  4170689372027579     fraud_Dare-Marvin   

             category     amt   first     last gender  \
491131      kids_pets  111.84    Jose  Vasquez      M   
491132      kids_pets   86.88     Ann   Lawson      F   
491133  entertainment   38.13  Samuel     Frey      M   

                            street          city state    zip      lat  \
491131         572 Davis Mountains  Lake Jackson    TX  77566  29.0393   
491132  144 Evans Islands Apt. 683       Burbank    WA  99323  46.1966   
491133    830 Myers Plaza Apt. 384        Edmond    OK  73034  35.6665   

            long  city_pop             job         dob  \
491131  -95.4401     28739  Futures trader  1999-12-27   
491132 -118.9017      3684        Musician  1981-11-29   
491133  -97.4798    116001     Media buyer  1993-05-10   

                               trans_num   unix_time  merch_lat  merch_long  \
491131  2090647dac2c89a1d86c514c427f5b91  1388534349  29.661049  -96.186633   
491132  6c5b7c8add471975aa0fec023b2e8408  1388534355  46.658340 -119.715054   
491133  1765bb45b3aa3224b4cdcb6e7a96cee3  1388534374  36.210097  -97.036372   

        is_fraud  
491131         0  
491132         0  
491133         0

In [7]:
cc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 491134 entries, 0 to 491133
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   trans_date_trans_time  491134 non-null  object 
 1   cc_num                 491134 non-null  int64  
 2   merchant               491134 non-null  object 
 3   category               491134 non-null  object 
 4   amt                    491134 non-null  float64
 5   first                  491134 non-null  object 
 6   last                   491134 non-null  object 
 7   gender                 491134 non-null  object 
 8   street                 491134 non-null  object 
 9   city                   491134 non-null  object 
 10  state                  491134 non-null  object 
 11  zip                    491134 non-null  int64  
 12  lat                    491134 non-null  float64
 13  long                   491134 non-null  float64
 14  city_pop               491134 non-nu

In [8]:
cc_df.describe()

cc_num            amt            zip            lat  \
count  4.911340e+05  491134.000000  491134.000000  491134.000000   
mean   3.706013e+17      69.050120   50770.532384      37.931230   
std    1.260229e+18     160.322867   26854.947965       5.341193   
min    5.038744e+11       1.000000    1843.000000      24.655700   
25%    2.131124e+14       8.960000   28405.000000      33.746700   
50%    3.531130e+15      42.170000   49628.000000      38.507200   
75%    4.653879e+15      80.330000   75048.000000      41.520500   
max    4.956829e+18   25086.940000   99323.000000      48.887800   

                long      city_pop     unix_time      merch_lat  \
count  491134.000000  4.911340e+05  4.911340e+05  491134.000000   
mean      -90.495619  1.213922e+05  1.358730e+09      37.930272   
std        12.990732  3.725751e+05  1.819402e+07       5.372986   
min      -122.345600  4.600000e+01  1.325376e+09      23.655789   
25%       -97.235100  1.228000e+03  1.343087e+09      33.781388   
50%       -87.591700  5.760000e+03  1.357257e+09      38.545124   
75%       -80.731000  5.083500e+04  1.374626e+09      41.624294   
max       -69.965600  2.906700e+06  1.388534e+09      49.887523   

          merch_long       is_fraud  
count  491134.000000  491134.000000  
mean      -90.495411       0.002533  
std        13.004100       0.050264  
min      -123.345106       0.000000  
25%       -96.984814       0.000000  
50%       -87.573441       0.000000  
75%       -80.685567       0.000000  
max       -68.965624       1.000000

# 2. 불필요한 컬럼 삭제 및 이름 변경

## 2-1. 불필요한 컬럼 삭제

- 개인 정보와 관련된 컬럼 삭제
- 대상 = [first, last, gender, street, city, state, zip, lat, long, dob]

In [9]:
pd.Series(cc_df.columns)

0     trans_date_trans_time
1                    cc_num
2                  merchant
3                  category
4                       amt
5                     first
6                      last
7                    gender
8                    street
9                      city
10                    state
11                      zip
12                      lat
13                     long
14                 city_pop
15                      job
16                      dob
17                trans_num
18                unix_time
19                merch_lat
20               merch_long
21                 is_fraud
dtype: object

In [10]:
toBeRemoved_col = cc_df.columns[[5, 6, 7, 8, 9, 10, 11, 12, 13, 16]]
cc_df.drop(toBeRemoved_col, axis=1, inplace=True)

In [11]:
cc_df.shape

(491134, 12)

## 2-2. 컬럼 이름 변경

- 신용카드 소유자 관련: holder_ 추가 
- 거래 상점 관련: merch_ 추가

In [12]:
pd.Series(cc_df.columns)

0     trans_date_trans_time
1                    cc_num
2                  merchant
3                  category
4                       amt
5                  city_pop
6                       job
7                 trans_num
8                 unix_time
9                 merch_lat
10               merch_long
11                 is_fraud
dtype: object

In [13]:
cc_df.rename({'trans_date_trans_time': 'trans_time', 'merchant':'merch_shop', 'category': 'merch_category', 'amt': 'amount', 'job':'holder_job'}, axis=1, inplace=True)

In [14]:
cc_df.head(2)

trans_time               cc_num  \
0  2019-01-01 00:00:44         630423337322   
1  2019-01-01 00:12:34  4956828990005111019   

                          merch_shop merch_category  amount  city_pop  \
0    fraud_Heller, Gutmann and Zieme    grocery_pos  107.23       149   
1  fraud_Schultz, Simonis and Little    grocery_pos   44.71    124967   

                          holder_job                         trans_num  \
0  Special educational needs teacher  1f76529f8574734946361c461b024d99   
1             Operational researcher  09eff9c806365e2a6be12c1bbab3d70e   

    unix_time  merch_lat  merch_long  is_fraud  
0  1325376044  49.159047 -118.186462         0  
1  1325376754  40.079588  -74.848087         0

# 3. 결측치 처리

- 결측치 없음
- is_fraud 컬럼에 0/1 이외 숫자 없음

In [15]:
cc_df.isna().any()

trans_time        False
cc_num            False
merch_shop        False
merch_category    False
amount            False
city_pop          False
holder_job        False
trans_num         False
unix_time         False
merch_lat         False
merch_long        False
is_fraud          False
dtype: bool

In [16]:
cc_df['is_fraud'].value_counts()

0    489890
1      1244
Name: is_fraud, dtype: int64

# 4. 이상치 처리 및 컬럼 값 타당성 검토

## 4-1. 신용카드 자릿수 (cc_num)

- 신용카드 자릿수는 14~16 범위로 이외 범위 삭제

In [17]:
cc_df['cc_num'] = cc_df['cc_num'].astype('str')

In [18]:
cc_df = cc_df[cc_df['cc_num'].str.len().between(14, 16)].reset_index(drop=True)

In [19]:
cc_df['cc_num'] = cc_df['cc_num'].astype('int')

In [20]:
cc_df.shape

(394078, 12)

## 4-2. 거래금액 (amount)

- 사기의심 거래금액의 범위는 1달러 ~ 25,086.94달러
- 로그 변환을 하여 극단적으로 큰 값과 작은 값의 스케일 차이를 줄여주어 데이터의 분포를 안정화시킴

In [21]:
cc_df['amount'].sort_values(ascending=False).head()

74465     25086.94
39623     17897.24
328755    16837.08
159236    15047.03
274084    12176.55
Name: amount, dtype: float64

In [22]:
cc_df['amount'].sort_values().head()

327992    1.0
299755    1.0
376298    1.0
314168    1.0
49192     1.0
Name: amount, dtype: float64

In [23]:
cc_df['amount'] = np.log(cc_df['amount'] )

In [24]:
cc_df.describe()

cc_num         amount      city_pop     unix_time      merch_lat  \
count  3.940780e+05  394078.000000  3.940780e+05  3.940780e+05  394078.000000   
mean   2.907809e+15       3.403577  1.049373e+05  1.358756e+09      37.929334   
std    2.127265e+15       1.417060  2.794766e+05  1.819502e+07       5.259905   
min    3.004433e+13       0.000000  4.600000e+01  1.325377e+09      25.421517   
25%    3.462732e+14       2.207175  1.725000e+03  1.343144e+09      33.748158   
50%    3.543591e+15       3.750915  6.006000e+03  1.357292e+09      38.347641   
75%    4.452366e+15       4.399867  5.428700e+04  1.374672e+09      41.739415   
max    6.538891e+15      10.130103  1.595797e+06  1.388534e+09      49.602783   

          merch_long       is_fraud  
count  394078.000000  394078.000000  
mean      -90.374975       0.002553  
std        13.169409       0.050461  
min      -123.345106       0.000000  
25%       -96.776345       0.000000  
50%       -88.274638       0.000000  
75%       -80.616977       0.000000  
max       -68.965624       1.000000

# 5. 중복 데이터 검토

- 중복된 데이터 없음

In [25]:
cc_df[cc_df.duplicated( )]

Empty DataFrame
Columns: [trans_time, cc_num, merch_shop, merch_category, amount, city_pop, holder_job, trans_num, unix_time, merch_lat, merch_long, is_fraud]
Index: []

# 6. 형태 변환

## 6-1. 거래시간 (trans_time)

- datetime으로 타입 변경

In [26]:
cc_df['trans_time'] = pd.to_datetime(cc_df['trans_time'])

In [27]:
cc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 394078 entries, 0 to 394077
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   trans_time      394078 non-null  datetime64[ns]
 1   cc_num          394078 non-null  int64         
 2   merch_shop      394078 non-null  object        
 3   merch_category  394078 non-null  object        
 4   amount          394078 non-null  float64       
 5   city_pop        394078 non-null  int64         
 6   holder_job      394078 non-null  object        
 7   trans_num       394078 non-null  object        
 8   unix_time       394078 non-null  int64         
 9   merch_lat       394078 non-null  float64       
 10  merch_long      394078 non-null  float64       
 11  is_fraud        394078 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(4), object(4)
memory usage: 36.1+ MB


## 6-2. 거래상점 (merch_shop)

- fradu_ 제거

In [28]:
cc_df['merch_shop'] = cc_df['merch_shop'].str.replace('fraud_', '')

## 6-3. 거래 품목 (merch_category)

- shopping, grocery, misc 항목은 매장 결제와 온라인 결제로 분리되어 있음
    - _pos: 매장에서 결제
    - _net: 온라인 결제
- 상품 카테고리별 분석만 수행하여 데이터 스케일을 줄여 분석 속도를 높이고자 함
- _pos, _net 삭제

In [29]:
cc_df['merch_category'].value_counts()

shopping_pos      39354
gas_transport     35552
home              34393
grocery_pos       33929
kids_pets         32376
shopping_net      32067
food_dining       29022
personal_care     28791
entertainment     27058
misc_pos          25989
health_fitness    25221
misc_net          20124
grocery_net       19965
travel            10237
Name: merch_category, dtype: int64

In [30]:
cc_df['merch_category'] = cc_df['merch_category'].str.replace(r'(_pos|_net)', '', regex=True)

In [31]:
cc_df['merch_category'].value_counts()

shopping          71421
grocery           53894
misc              46113
gas_transport     35552
home              34393
kids_pets         32376
food_dining       29022
personal_care     28791
entertainment     27058
health_fitness    25221
travel            10237
Name: merch_category, dtype: int64

# 7. 새로운 컬럼 추가 (feature engineering)

## 7-1. 거래 시간대 컬럼

- 거래 시간을 morning, afternoon, evening, night로 구분
- 특정 시간대에 신용카드 이상거래가 빈번히 발생할 수 있음
- 거래 시간을 구분하면 모델이 이러한 패턴을 학습하는 데 도움이 될 것임

In [32]:
def assign_timeOfDay(datetime):
    hour = datetime.hour
    if 6 <= hour < 12:
        return 'morning'
    elif 12 <= hour < 18:
        return 'afternoon'
    elif 18 <= hour < 24:
        return 'evening'
    else:
        return 'night'

In [33]:
cc_df['trans_timeOfDay'] = cc_df['trans_time'].apply(assign_timeOfDay)

In [34]:
cc_df['trans_timeOfDay'].value_counts()

afternoon    142452
evening      141715
morning       54999
night         54912
Name: trans_timeOfDay, dtype: int64

## 7-2. 거래 금액의 Z-score 컬럼

- 정상 거래(is_fraud == 0)에 한하여, 거래 금액의 평균, 표준편차를 구한뒤,
- 모든 거래 금액의 Z-score를 계산하여 컬럼을 만듬
- Z-score는 특정 거래가 평균 거래 금액에서 얼마나 벗어나 있는지를 직관적으로 이해하게 하며, 모델이 이러한 패턴을 학습하는데 도움이 될 것임

In [35]:
cc_df['is_fraud'].value_counts()

0    393072
1      1006
Name: is_fraud, dtype: int64

In [36]:
trans_normal = cc_df[cc_df['is_fraud'] == 0]

In [37]:
trans_mean = trans_normal['amount'].mean()
trans_std = trans_normal['amount'].std()

In [38]:
cc_df['amount_zscore'] = (cc_df['amount'] - trans_mean) / trans_std

In [39]:
cc_df.head(2)

trans_time           cc_num       merch_shop merch_category  \
0 2019-01-01 00:17:16  180048185037117      Kling-Grant        grocery   
1 2019-01-01 00:20:15  374930071163758  Deckow-O'Conner        grocery   

     amount  city_pop              holder_job  \
0  3.834710     71485  Leisure centre manager   
1  4.160288     31515          Police officer   

                          trans_num   unix_time  merch_lat  merch_long  \
0  19e23c6a300c774354417befe4f31f8c  1325377036  40.021888  -74.228188   
1  6f363661ba6b55889e488dd178f2a0af  1325377215  42.360426  -83.552316   

   is_fraud trans_timeOfDay  amount_zscore  
0         0           night       0.308584  
1         0           night       0.539018

## 7-3. 인구수 범주화 컬럼

- small_town (인구 < 10만), medium_sized_city (10만 ≤ 인구 < 100만), large_city (인구 ≥ 100만)로 범주화

In [40]:
def assign_population(population):
    if population < 100000:
        return 'small_town'
    elif 100000 <= population < 1000000:
        return 'medium_sized_city'
    else:
        return 'large_city'

In [41]:
cc_df['popula_category'] = cc_df['city_pop'].apply(assign_population)

In [42]:
cc_df['popula_category'].value_counts()

small_town           316006
medium_sized_city     65676
large_city            12396
Name: popula_category, dtype: int64

## 7-4. 인구수 대비 거래금액 컬럼

- 특정 거래금액이 해당 지역의 경제 규모에 비해 과도한지 여부를 판단하는 데 도움
- 금액의 크기가 클수록 1인당 부담이 커지므로, 이상치 포착에 유리

In [43]:
cc_df['pop_amount_ratio'] = cc_df['amount'] / cc_df['city_pop']

In [44]:
cc_df.head(3)

trans_time            cc_num        merch_shop merch_category  \
0 2019-01-01 00:17:16   180048185037117       Kling-Grant        grocery   
1 2019-01-01 00:20:15   374930071163758   Deckow-O'Conner        grocery   
2 2019-01-01 00:23:41  2712209726293386  Balistreri-Nader           misc   

     amount  city_pop              holder_job  \
0  3.834710     71485  Leisure centre manager   
1  4.160288     31515          Police officer   
2  3.241811    378909     Designer, furniture   

                          trans_num   unix_time  merch_lat  merch_long  \
0  19e23c6a300c774354417befe4f31f8c  1325377036  40.021888  -74.228188   
1  6f363661ba6b55889e488dd178f2a0af  1325377215  42.360426  -83.552316   
2  1654da2abfb9e79a5f99167fc9779558  1325377421  29.737426  -90.853194   

   is_fraud trans_timeOfDay  amount_zscore    popula_category  \
0         0           night       0.308584         small_town   
1         0           night       0.539018         small_town   
2         0           night      -0.111051  medium_sized_city   

   pop_amount_ratio  
0          0.000054  
1          0.000132  
2          0.000009

## 7-5. 거래 발생 주기 컬럼

- 동일한 카드로 발생한 이전 거래와의 시간 간격을 계산하여 비정상적으로 짧은 시간 내 반복되는 거래를 탐지
- 임계값(threshold)은 1시간으로 설정
- 동일한 카드로 임계값 이내 다시 거래가 있을 경우 1로 표시, 아닐 경우 0으로 표시
- 임계값 설정의 타당성은 검토하지 않음

In [45]:
# NaT 값 주의!!
cc_df['time_since_last_transaction'] = cc_df.groupby('cc_num')['trans_time'].diff()

In [46]:
threshold = pd.Timedelta(hours=1)
cc_df['short_interval_trans'] = cc_df['time_since_last_transaction'].apply(lambda x: 1 if x > threshold else 0 if x < threshold else 0)

In [47]:
cc_df['short_interval_trans'].value_counts()

1    288144
0    105934
Name: short_interval_trans, dtype: int64

In [48]:
cc_df.head()

trans_time            cc_num        merch_shop merch_category  \
0 2019-01-01 00:17:16   180048185037117       Kling-Grant        grocery   
1 2019-01-01 00:20:15   374930071163758   Deckow-O'Conner        grocery   
2 2019-01-01 00:23:41  2712209726293386  Balistreri-Nader           misc   
3 2019-01-01 00:26:22  3598215285024754   Heidenreich PLC        grocery   
4 2019-01-01 00:27:45   213141712584544   Halvorson Group           misc   

     amount  city_pop              holder_job  \
0  3.834710     71485  Leisure centre manager   
1  4.160288     31515          Police officer   
2  3.241811    378909     Designer, furniture   
3  5.334456     34496       Librarian, public   
4  5.200429     19685             Fine artist   

                          trans_num   unix_time  merch_lat  merch_long  \
0  19e23c6a300c774354417befe4f31f8c  1325377036  40.021888  -74.228188   
1  6f363661ba6b55889e488dd178f2a0af  1325377215  42.360426  -83.552316   
2  1654da2abfb9e79a5f99167fc9779558  1325377421  29.737426  -90.853194   
3  f246ac27c09429ff1e385d4683c79c4d  1325377582  40.202538  -73.859212   
4  7bb25a43205191eb7344282b88fc54d3  1325377665  31.939812  -90.726209   

   is_fraud trans_timeOfDay  amount_zscore    popula_category  \
0         0           night       0.308584         small_town   
1         0           night       0.539018         small_town   
2         0           night      -0.111051  medium_sized_city   
3         0           night       1.370057         small_town   
4         0           night       1.275197         small_town   

   pop_amount_ratio time_since_last_transaction  short_interval_trans  
0          0.000054                         NaT                     0  
1          0.000132                         NaT                     0  
2          0.000009                         NaT                     0  
3          0.000155                         NaT                     0  
4          0.000264                         NaT                     0

# 8. 불필요한 컬럼 삭제

## 8-1. feature engineering으로 분석 완료한 컬럼 삭제

In [49]:
cc_df.drop( [ 'trans_time', 'cc_num', 'amount', 'city_pop', 'time_since_last_transaction' ],  axis=1, inplace=True)

## 8-2. 머신러닝 알고리즘에 적용하지 않을 컬럼 삭제

- merch_category 컬럼을 제외하고 모두 삭제
- 삭제 이유: 유니크한 개수가 많아 패턴을 분석하기 어렵고, 다른 컬럼들과 유의미한 분석을 진행하지 않음
- 삭제되는 컬럼들에 대해 좀더 유의미한 분석을 하지 못한 아쉬움!!!

In [50]:
pd.Series(cc_df.columns)

0               merch_shop
1           merch_category
2               holder_job
3                trans_num
4                unix_time
5                merch_lat
6               merch_long
7                 is_fraud
8          trans_timeOfDay
9            amount_zscore
10         popula_category
11        pop_amount_ratio
12    short_interval_trans
dtype: object

In [51]:
def count_unique(col):
    print(f'\n* {col}: {cc_df[col].nunique()}개')

In [52]:
deleteOrNot_cols = cc_df.columns[[0, 1, 2, 3, 4, 5, 6]]

print('== 유니크한 개수==')
for col in deleteOrNot_cols:
    count_unique(col)

== 유니크한 개수==

* merch_shop: 693개

* merch_category: 11개

* holder_job: 93개

* trans_num: 394078개

* unix_time: 392340개

* merch_lat: 389722개

* merch_long: 391717개


In [53]:
cc_df.drop( [ 'merch_shop', 'holder_job', 'trans_num', 'unix_time', 'merch_lat', 'merch_long' ],  axis=1, inplace=True)

In [54]:
cc_df.shape

(394078, 7)

In [55]:
cc_df.head()

merch_category  is_fraud trans_timeOfDay  amount_zscore    popula_category  \
0        grocery         0           night       0.308584         small_town   
1        grocery         0           night       0.539018         small_town   
2           misc         0           night      -0.111051  medium_sized_city   
3        grocery         0           night       1.370057         small_town   
4           misc         0           night       1.275197         small_town   

   pop_amount_ratio  short_interval_trans  
0          0.000054                     0  
1          0.000132                     0  
2          0.000009                     0  
3          0.000155                     0  
4          0.000264                     0

In [56]:
cc_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 394078 entries, 0 to 394077
Data columns (total 7 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   merch_category        394078 non-null  object 
 1   is_fraud              394078 non-null  int64  
 2   trans_timeOfDay       394078 non-null  object 
 3   amount_zscore         394078 non-null  float64
 4   popula_category       394078 non-null  object 
 5   pop_amount_ratio      394078 non-null  float64
 6   short_interval_trans  394078 non-null  int64  
dtypes: float64(2), int64(2), object(3)
memory usage: 21.0+ MB


# 9. One-hot Encoding

In [57]:
encoded_cc_df = pd.get_dummies(cc_df, columns = ['merch_category','trans_timeOfDay','popula_category'], drop_first = True)

In [58]:
encoded_cc_df.head()

is_fraud  amount_zscore  pop_amount_ratio  short_interval_trans  \
0         0       0.308584          0.000054                     0   
1         0       0.539018          0.000132                     0   
2         0      -0.111051          0.000009                     0   
3         0       1.370057          0.000155                     0   
4         0       1.275197          0.000264                     0   

   merch_category_food_dining  merch_category_gas_transport  \
0                           0                             0   
1                           0                             0   
2                           0                             0   
3                           0                             0   
4                           0                             0   

   merch_category_grocery  merch_category_health_fitness  merch_category_home  \
0                       1                              0                    0   
1                       1                              0                    0   
2                       0                              0                    0   
3                       1                              0                    0   
4                       0                              0                    0   

   merch_category_kids_pets  merch_category_misc  \
0                         0                    0   
1                         0                    0   
2                         0                    1   
3                         0                    0   
4                         0                    1   

   merch_category_personal_care  merch_category_shopping  \
0                             0                        0   
1                             0                        0   
2                             0                        0   
3                             0                        0   
4                             0                        0   

   merch_category_travel  trans_timeOfDay_evening  trans_timeOfDay_morning  \
0                      0                        0                        0   
1                      0                        0                        0   
2                      0                        0                        0   
3                      0                        0                        0   
4                      0                        0                        0   

   trans_timeOfDay_night  popula_category_medium_sized_city  \
0                      1                                  0   
1                      1                                  0   
2                      1                                  1   
3                      1                                  0   
4                      1                                  0   

   popula_category_small_town  
0                           1  
1                           1  
2                           0  
3                           1  
4                           1

# [회고]

### 1. 개인정보관련 삭제
- 개인정보를 활용하면 더 많은 분석이 가능할 것임
- 예를 들어 거주지와 거래 장소의 거리 계산을 통해 너무 먼 거리일 경우 이상 거래일 가능성이 높을 것임
- 개인정보를 활용하지 않고 분석할 수 있는 방법에 대해 생각해 봄
- 현업에서는 개인정보를 어느정도 활용하는지 궁금하며, 법적으로 어떤 제한이 있는지 궁금해짐

### 2. 거래 발생 주기 컬럼에서 임계값 설정
- 임계값을 1시간으로 설정했으나, 타당한 설정인지 판단되지 않음
- 데이터 전처리 단계에서 임계값 설정에 대하여 고민해봄

### 3. 카드 소유자 직업 범주화
- 특정 직업군에서 사기 거래가 더 많이 발생하는지에 대한 분석은 중요함
- 카드 소유자 직업의 유니크한 개수가 93개로 범주화할 필요가 있으나 범주화 기준에 대해 적절한 방법을 찾지 못함
- 인터넷 서치를 통해 알아내야 하나, 시간 제약으로 하지 못함

### 4. 머신러닝 알고리즘에 사용하지 않을 컬럼 삭제
- 데이터 셋에는 많은 양의 정보가 담겨 있으나, 충분히 분석하지 못해 아쉬움이 있음